### PR TITLE
[SPARK-33407][PYTHON] Simplify the exception message from Python UDFs (disabled by default)

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to simplify the exception messages from Python UDFS. 

Currently, the exception message from Python UDFs is as below:

```python
from pyspark.sql.functions import udf; spark.range(10).select(udf(lambda x: x/0)("id")).collect()
```

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../python/pyspark/sql/dataframe.py", line 427, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/.../python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1305, in __call__
  File "/.../python/pyspark/sql/utils.py", line 127, in deco
    raise_from(converted)
  File "<string>", line 3, in raise_from
pyspark.sql.utils.PythonException:
  An exception was thrown from Python worker in the executor:
Traceback (most recent call last):
  File "/.../python/lib/pyspark.zip/pyspark/worker.py", line 605, in main
    process()
  File "/.../python/lib/pyspark.zip/pyspark/worker.py", line 597, in process
    serializer.dump_stream(out_iter, outfile)
  File "/.../python/lib/pyspark.zip/pyspark/serializers.py", line 223, in dump_stream
    self.serializer.dump_stream(self._batched(iterator), stream)
  File "/.../python/lib/pyspark.zip/pyspark/serializers.py", line 141, in dump_stream
    for obj in iterator:
  File "/.../python/lib/pyspark.zip/pyspark/serializers.py", line 212, in _batched
    for item in iterator:
  File "/.../python/lib/pyspark.zip/pyspark/worker.py", line 450, in mapper
    result = tuple(f(*[a[o] for o in arg_offsets]) for (arg_offsets, f) in udfs)
  File "/.../python/lib/pyspark.zip/pyspark/worker.py", line 450, in <genexpr>
    result = tuple(f(*[a[o] for o in arg_offsets]) for (arg_offsets, f) in udfs)
  File "/.../python/lib/pyspark.zip/pyspark/worker.py", line 90, in <lambda>
    return lambda *a: f(*a)
  File "/.../python/lib/pyspark.zip/pyspark/util.py", line 107, in wrapper
    return f(*args, **kwargs)
  File "<stdin>", line 1, in <lambda>
ZeroDivisionError: division by zero
```

Actually, almost all cases, users only care about `ZeroDivisionError: division by zero`. We don't really have to show the internal stuff in 99% cases. 

This PR adds a configuration `spark.sql.execution.pyspark.udf.simplifiedException.enabled` (disabled by default) that hides the internal tracebacks related to Python worker, (de)serialization, etc.

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../python/pyspark/sql/dataframe.py", line 427, in show
    print(self._jdf.showString(n, 20, vertical))
  File "/.../python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1305, in __call__
  File "/.../python/pyspark/sql/utils.py", line 127, in deco
    raise_from(converted)
  File "<string>", line 3, in raise_from
pyspark.sql.utils.PythonException:
  An exception was thrown from Python worker in the executor:
Traceback (most recent call last):
  File "<stdin>", line 1, in <lambda>
ZeroDivisionError: division by zero
```

The trackback will be shown from the point when any non-PySpark file is seen in the traceback.

### Why are the changes needed?

Without this configuration. such internal tracebacks are exposed to users directly especially for shall or notebook users in PySpark. 99% cases people don't care about the internal Python worker, (de)serialization and related tracebacks. It just makes the exception more difficult to read. For example, one statement of `x/0` above shows a very long traceback and most of them are unnecessary. 

This configuration enables the ability to show simplified tracebacks which users will likely be most interested in.

### Does this PR introduce _any_ user-facing change?

By default, no. It adds one configuration that simplifies the exception message. See the example above.

### How was this patch tested?

Manually tested:

```bash
$ pyspark --conf spark.sql.execution.pyspark.udf.simplifiedException.enabled=true
```
```python
from pyspark.sql.functions import udf; spark.sparkContext.setLogLevel("FATAL"); spark.range(10).select(udf(lambda x: x/0)("id")).collect()
```

and unittests were also added.

### CODE REVIEW

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -604,17 +604,19 @@ def process():
...
         TaskContext._setTaskContext(None)
         BarrierTaskContext._setTaskContext(None)
-    except BaseException:
+    except BaseException as e:
         try:
-            exc_info = traceback.format_exc()
-            if isinstance(exc_info, bytes):
-                # exc_info may contains other encoding bytes, replace the invalid bytes and convert
-                # it back to utf-8 again
-                exc_info = exc_info.decode("utf-8", "replace").encode("utf-8")
```
Comment (HyukjinKwon) [inline python/pyspark/worker.py:613]: We dropped Python 2. It doesn't need it anymore.

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/util.py [hunk 2]
```diff
@@ -75,6 +79,144 @@ def wrapper(*args, **kwargs):
...
+    -------
+    TracebackType or None
+      Simplified traceback instance. It returns None if it fails to simplify.
+
+    Notes
+    -----
+    This keeps the tracebacks once it sees they are from a different file even
+    though the following tracebacks are from the current package.
+
+    Examples
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:103]: I thought it's easier to understand this traceback modification with the doctests. But let me know if any of you think it should be converted and placed as a proper unittest under `test_xxx.py`.

**Comment by HyukjinKwon** [comment]
This is ready for a review. I tested in notebooks as well.

**Comment by HyukjinKwon** [comment]
cc @BryanCutler, @viirya, @ueshin and @srowen FYI.

**Comment by HyukjinKwon** [comment]
cc @zero323 @BryanCutler, @viirya, @ueshin and @srowen FYI.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -1927,6 +1927,16 @@ object SQLConf {
       .version("3.0.0")
       .fallbackConf(BUFFER_SIZE)
 
+  val PYSPARK_SIMPLIFIEID_EXCEPTION =
+    buildConf("spark.sql.execution.pyspark.udf.simplifiedException.enabled")
+      .doc(
+        s"When true, the exception messages from Python UDFs are simplified. It hides " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1933]: nit: s"" -> ""?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -1927,6 +1927,16 @@ object SQLConf {
       .version("3.0.0")
       .fallbackConf(BUFFER_SIZE)
 
+  val PYSPARK_SIMPLIFIEID_EXCEPTION =
+    buildConf("spark.sql.execution.pyspark.udf.simplifiedException.enabled")
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1931]: simplifiedException or simplifiedTraceback?

File: python/pyspark/util.py [hunk 5]
```diff
@@ -20,6 +20,10 @@
 import re
 import sys
 import traceback
+import types
+import os
+import platform
+import itertools
```
Comment (BryanCutler) [inline python/pyspark/util.py:26]: nit: imports not in alphabetical order - not necessary though

**Review by viirya — COMMENTED** [review]
Looks a good idea.

File: python/pyspark/util.py [hunk 6]
```diff
@@ -75,6 +79,144 @@ def wrapper(*args, **kwargs):
...
+    ...     pairs = zip(walk_tb(tb), traceback.extract_tb(tb))
+    ...     for cur_tb, cur_frame in pairs:
+    ...         if cur_frame.filename.startswith(root):
+    ...             return cur_tb
+
+    Regular exceptions should show the file name of the current package as below.
+
+    >>> exc_info = None
+    >>> try:
+    ...    fail_on_stopiteration(dummy_module.raise_stop_iteration)()
```
Comment (BryanCutler) [inline python/pyspark/util.py:135]: indentation off here?

File: python/pyspark/util.py [hunk 7]
```diff
@@ -75,6 +79,144 @@ def wrapper(*args, **kwargs):
...
+    Traceback (most recent call last):
+      File ...
+        ...
+      File "/.../pyspark/util.py", line ...
+        ...
+    RuntimeError: ...
+    >>> "pyspark/util.py" in exc_info
+    True
+
+    If the the traceback is simplified with this method, it hides the current package file name:
```
Comment (BryanCutler) [inline python/pyspark/util.py:151]: "the the" -> "the"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -1927,6 +1927,16 @@ object SQLConf {
...
+      .doc(
+        s"When true, the exception messages from Python UDFs are simplified. It hides " +
+         "the Python worker, (de)serialization, etc from PySpark in tracebacks, and only" +
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1934]: need a space after 'only'?

File: python/pyspark/util.py [hunk 9]
```diff
@@ -177,6 +319,8 @@ def __del__(self):
 
 if __name__ == "__main__":
     import doctest
-    (failure_count, test_count) = doctest.testmod()
-    if failure_count:
-        sys.exit(-1)
+
+    if "pypy" not in platform.python_implementation().lower() and sys.version_info[:2] >= (3, 7):
+        (failure_count, test_count) = doctest.testmod()
```
Comment (viirya) [inline python/pyspark/util.py:324]: Does this mean we also skip other doctest in `python/pyspark/util.py`?

File: python/pyspark/util.py (hunk 6 above)
Comment (BryanCutler) [reply python/pyspark/util.py:135]: also looks like similar spots below.

**Review by BryanCutler — COMMENTED** [review]
LGTM, simplified messages will be nice to have. I only had a few minor nits.

File: python/pyspark/util.py [hunk 10]
```diff
@@ -183,6 +325,8 @@ def __del__(self):
 
 if __name__ == "__main__":
     import doctest
-    (failure_count, test_count) = doctest.testmod()
-    if failure_count:
-        sys.exit(-1)
+
+    if "pypy" not in platform.python_implementation().lower() and sys.version_info[:2] >= (3, 7):
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:329]: @viirya, let me just skip the tests here. Seems like it's tricky to conditionally skip the tests.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Let me merge this actually. GitHub Actions build pass and all different PyPy/Python versions are tested.
Thanks for approval and reviews, @viirya and @BryanCutler.

Merged to master.

**Pull request closed without merging**